### DISCUSSION TITLE
Possible in auto type conversion in the Literal field ?

### DISCUSSION THREAD

**Discussion by zhfkt** [pydantic/pydantic#3778] [Question]
Hi,

I am wondering whether pydantic support in auto type conversion in the Literal field ?  

```python


from typing_extensions import Literal

from pydantic import BaseModel

class TestClass(BaseModel):
    value: int


print(TestClass(value=1))
print(TestClass(value="1"))


class TestLiteralClass(BaseModel):
    value: Literal[1, 2, 3]


print(TestLiteralClass(value=1))
print(TestLiteralClass(value="1"))


```

Output:

```text

value=1
value=1
value=1
Traceback (most recent call last):
  File "temp.py", line 19, in <module>
    print(TestLiteralClass(value="1"))
  File "pydantic\main.py", line 406, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for TestLiteralClass
value
  unexpected value; permitted: 1, 2, 3 (type=value_error.const; given=1; permitted=(1, 2, 3))


```

We could give the type integer and str to the value in `TestClass` , which supports converting str to int. However, `TestLiteralClass` will report error and it doesn't seem support converting str to int.  I am asking for help and not sure whether we could have a workaround to achieve auto type conversion in the Literal field.

Thank you !
Best Regards

**Comment by itsayellow** [comment]
I'm wondering the exact same thing.

In my case I'm interested in using a discriminator to decide between two different models in a union.  Specifying the discriminator field requires that that field is typed with a Literal.  However, when using a Literal it seems that you give up any automatic type conversion.  

In my case I am receiving json data where numbers are given as strings, and I'd really love to both specify that model1 wants field1 to be 2, while model2 wants field1 to be 0 or 1.  I'd like them to be int's in the model instance.  But I'd still like to accept an input that specifies `field1="0"`!

I 
This is what I'm trying to do, that results in an error:
```py
from typing import Annotated, List
                                                                                
from pydantic import BaseModel, Field                                           

class MyModel_small(BaseModel):
    field1: int = Field(le=1)
class MyModel_big(BaseModel):
    field1: int = Field(ge=2,le=2)

MyModel_all = Annotated[MyModel_small|MyModel_big, Field(discriminator="field1")]

class ParentModel(BaseModel):
    mixed_list: List[MyModel_all]

new_model = ParentModel(mixed_list=[{"field1":"2"},{"field1":"1"}])
```
Output is pages of errors ending with:
```
  File "[...]/venv311/lib/python3.11/site-packages/pydantic/_internal/_discriminated_union.py", line 488, in _infer_discriminator_values_for_inner_schema
    raise PydanticUserError(
pydantic.errors.PydanticUserError: Model 'MyModel_small' needs field 'field1' to be of type `Literal`

For further information visit https://errors.pydantic.dev/2.5/u/discriminator-needs-literal
```

Mixed Lists like this seem to work without a discriminator, but I was hoping to add a discriminator to make things more efficient.

**Comment by EliotGodard** [comment]
Same problem here with this piece of code. I had to write a custom validator. I was wondering if there was a type annotation that could be used to automatically coerce the value.

```Python
class MySchema(BaseModel):
    my_var: Literal[2, 10, 15] = 10

    # Coerce my_var because Pydantic doesn't automaticaly do it for Literal
    @field_validator("my_var ", mode="before")
    @classmethod
    def coerce_my_var _to_int(cls, v: Any) -> int:
        if isinstance(v, str):
            try:
                int(v)
            except Exception as err:
                raise ValueError("my_var must be an int") from err
        return int(v)
```

**Reply by EliotGodard** [reply]
It's also possible to use an IntEnum but again, seems overcomplicated for what it's trying to do.

```Python
class MyEnum(IntEnum):
    _2 = 2
    _10 = 10
    _15 = 15


class SchemaRequeteBaseEcs(BaseModel):
    my_var: MyEnum = MyEnum._10
```

**Reply by sarthak7gupta** [reply]
Hi, I know I'm super late, but ran into the same issue and found this thread.

I ended up using Annotated with a BeforeValidator to make the coercion a bit cleaner and avoid writing a custom validator method.

```py
from typing import Annotated, Literal
from pydantic import BaseModel, BeforeValidator

class MySchema(BaseModel):
    my_var: Annotated[Literal[2, 10, 15], BeforeValidator(int)] = 10
    my_optional_var: Annotated[Literal[2, 10, 15], BeforeValidator(int)] | None = None
```
    
This keeps the model declarative and avoids the extra boilerplate of a manual validator. Hope it helps someone else who stumbles across this!    

Note: In case the field is optional, don’t include None inside the Annotated, otherwise the BeforeValidator will try to coerce None to an int.
`Annotated[Literal[2, 10, 15] | None, BeforeValidator(int)] | None` or  `Annotated[Literal[2, 10, 15, None], BeforeValidator(int)] | None` won't work

**Comment by Viicos** [comment]
This is tracked in https://github.com/pydantic/pydantic/issues/9991.
